### PR TITLE
Add qrio: Decode QR codes with quirc lib

### PR DESCRIPTION
**Pull request by jepler**
Camera-free testing with qr.ppm from [qr.zip](https://github.com/adafruit/circuitpython/files/6931916/qr.zip):
```
import qrio
with open("qr.pgm", "rb") as f: content = f.read()[-76800:]
d = qrio.QRDecoder(320,240, qrio.PixelPolicy.EVERY_BYTE)
for row in d.decode(content):
    print(row)
```

Also works with OV2640: [code.py.txt](https://github.com/adafruit/circuitpython/files/6931931/code.py.txt)

This pulls in the quirc library, an adafruit fork of https://github.com/dlbeer/quirc/ -- Its license:

> quirc -- QR-code recognition library
> Copyright (C) 2010-2012 Daniel Beer <dlbeer@gmail.com>
> 
> ISC License
> ===========
> 
> Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee is hereby granted, provided that the above copyright notice and this permission notice appear in all copies.
> 
> THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

### CODE REVIEW

**Comment by jerryneedell** [comment]
@jepler Very cool! I tried this on my Kaluga v1.3 with your example. Works well as lomg as the qr code image was big enough!
BTW, I think the example usage of the button is not working properly, it did not seem to wait for the button press, so I removed it rather then troubleshoot. 
```
Adafruit CircuitPython 7.0.0-alpha.5-98-gbd903f2f0 on 2021-08-04; Kaluga 1 with ESP32S2
>>> import qr
a
a
a
b'http://s.espruino.com'
b'http://s.espruino.com'
b'http://s.espruino.com'
b'http://s.espruino.com'
b'http://s.espruino.com'
b'http://s.espruino.com'
b'http://s.espruino.com'
b'http://s.espruino.com'
```

File: ports/esp32s2/mpconfigport.mk [hunk 1]
```diff
@@ -31,7 +31,7 @@ CIRCUITPY_ROTARYIO = 1
 CIRCUITPY_NVM = 1
 CIRCUITPY_PS2IO ?= 1
 CIRCUITPY_TOUCHIO_USE_NATIVE ?= 1
-CIRCUITPY_WIFI = 1
+CIRCUITPY_WIFI ?= 1
```
Comment (tannewt) [inline ports/esp32s2/mpconfigport.mk:34]: What does this do?

File: shared-bindings/qrio/QRDecoder.c [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+#include "shared-module/qrio/QRDecoder.h"
+
+#include "py/obj.h"
+#include "py/objproperty.h"
+#include "py/enum.h"
+
+//| class QRDecoder:
+//|
+//|     def __init__(self, width: int, height: int) -> None:
+//|         """Construct a QRDecoder object"""
```
Comment (tannewt) [inline shared-bindings/qrio/QRDecoder.c:38]: How are width and height used?

File: shared-bindings/qrio/QRDecoder.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "py/obj.h"
+
+typedef struct qrio_qrdecoder_obj qrio_qrdecoder_obj_t;
+
+extern const mp_obj_type_t qrio_qrdecoder_type_obj;
+
+void common_hal_qrio_qrdecoder_construct(qrio_qrdecoder_obj_t *self);
+
+mp_obj_t common_hal_qrio_qrdecoder_recognize(qrio_qrdecoder_obj_t *self, int width, int height, mp_buffer_info_t *buf);
```
Comment (tannewt) [inline shared-bindings/qrio/QRDecoder.h:37]: Should this be decode instead of recognize?

File: shared-bindings/qrio/__init__.c [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    MAKE_ENUM_MAP_ENTRY(qrio_pixel_policy, EVEN_BYTES),
+    MAKE_ENUM_MAP_ENTRY(qrio_pixel_policy, ODD_BYTES),
+};
+STATIC MP_DEFINE_CONST_DICT(qrio_pixel_policy_locals_dict, qrio_pixel_policy_locals_table);
+
+MAKE_PRINTER(qrio, qrio_pixel_policy);
+
+MAKE_ENUM_TYPE(qrio, PixelPolicy, qrio_pixel_policy);
+
+//| class qrinfo:
```
Comment (tannewt) [inline shared-bindings/qrio/__init__.c:63]: Why lower case for the name? I'd prefer this as a separate file so it's easier to find.

**Review by tannewt — CHANGES_REQUESTED** [review]
Just a few questions. Nice work on this!

File: shared-bindings/qrio/QRDecoder.c (hunk 2 above)
Comment (jepler) [reply shared-bindings/qrio/QRDecoder.c:38]: They set the width and height of a buffer internal to the quirc library

File: ports/esp32s2/mpconfigport.mk (hunk 1 above)
Comment (jepler) [reply ports/esp32s2/mpconfigport.mk:34]: It means I can place "CIRCUITPY_WIFI = 0" in `makefile` (all lowercase) to override the setting.  Disabling wifi is usually needed to make debug builds fit.

File: shared-bindings/qrio/__init__.c (hunk 4 above)
Comment (jepler) [reply shared-bindings/qrio/__init__.c:63]: OK.

File: shared-bindings/qrio/QRDecoder.c (hunk 2 above)
Comment (tannewt) [reply shared-bindings/qrio/QRDecoder.c:38]: Please say so in the docs.

Comment (jepler) [reply shared-bindings/qrio/QRDecoder.c]: done

File: shared-bindings/qrio/QRDecoder.h (hunk 3 above)
Comment (jepler) [reply shared-bindings/qrio/QRDecoder.h]: done

File: shared-bindings/qrio/__init__.c (hunk 4 above)
Comment (jepler) [reply shared-bindings/qrio/__init__.c]: done.

**Comment by jepler** [comment]
@tannewt ready for your re-review!

**Review by tannewt — APPROVED** [review]
Looks good! Thank you for the extra docs and reorg.

**Pull request merged**